nestkernel/sparse_node_array.cpp

```diff
@@ -24,81 +24,114 @@
 
 // Includes from nestkernel:
 #include "exceptions.h"
+#include "kernel_manager.h"
 #include "node.h"
+#include "vp_manager_impl.h"
+
+
+#include <iomanip>
+#include <iostream>
 
 nest::SparseNodeArray::NodeEntry::NodeEntry( Node& node, index node_id )
   : node_( &node )
   , node_id_( node_id )
 {
-  assert( node_id == node.get_node_id() );
 }
 
 nest::SparseNodeArray::SparseNodeArray()
   : nodes_()
-  , max_node_id_( 0 )
+  , global_max_node_id_( 0 )
   , local_min_node_id_( 0 )
   , local_max_node_id_( 0 )
-  , node_id_idx_scale_( 1.0 )
+  , left_scale_( 1.0 )
+  , right_scale_( 1.0 )
+  , split_node_id_( 0 )
+  , split_idx_( 0 )
+  , have_split_( false )
+  , left_side_has_proxies_( false ) // meaningless initial value
 {
 }
 
+void
+nest::SparseNodeArray::clear()
+{
+  nodes_.clear();
+
+  global_max_node_id_ = 0;
+  local_min_node_id_ = 0;
+  local_max_node_id_ = 0;
+  left_scale_ = 1.0;
+  right_scale_ = 1.0;
+  split_node_id_ = 0;
+  split_idx_ = 0;
+  have_split_ = false;
+  left_side_has_proxies_ = false;
+}
+
 void
 nest::SparseNodeArray::add_local_node( Node& node )
 {
   const index node_id = node.get_node_id();
 
-  // protect against node ID 0
-  assert( node_id > 0 );
-
-  // local_min_node_id_ can only be 0 if no node has been stored
-  assert( local_min_node_id_ > 0 or nodes_.size() == 0 );
+  // ensure increasing order
+  assert( node_id > local_max_node_id_ );
 
-  // local_min_node_id_ cannot be larger than local_max_node_id_
-  assert( local_min_node_id_ <= local_max_node_id_ );
-
-  // local_max_node_id_ cannot be larger than max_node_id_
-  assert( local_max_node_id_ <= max_node_id_ );
+  nodes_.push_back( NodeEntry( node, node_id ) );
+  local_max_node_id_ = node_id;
 
-  // node_id must exceed max_node_id_
-  assert( node_id > max_node_id_ );
+  // mark array inconsistent until set_max_node_id() called
+  global_max_node_id_ = 0;
 
-  // all is consistent, register node and update auxiliary variables
-  nodes_.push_back( NodeEntry( node, node_id ) );
-  if ( local_min_node_id_ == 0 ) // only first non-zero
+  // Setup when first node is added
```

Comment:
All other comments here start with a lower case
```suggestion
  // setup when first node is added
```

Reply:
OK.